openmls/src/group/public_group/validation.rs

```diff
@@ -308,6 +309,75 @@ impl PublicGroup {
         Ok(encryption_keys)
     }
 
+    /// Validate PreSharedKey proposals. This function implements the following checks:
+    /// // TODO: ValSem
+    /// * PSK in proposal must be of type Resumption/Application or External.
+    /// * The application SHOULD specify an upper limit on the number of past epochs for which the resumption_psk may be stored.
+    /// * It contains multiple PreSharedKey proposals that reference the same PreSharedKeyID.
+    /// * Must not contain multiple PreSharedKey proposals that reference the same PreSharedKeyID.
+    pub(crate) fn validate_pre_shared_key_proposals(
+        &self,
+        proposal_queue: &ProposalQueue,
+    ) -> Result<(), ProposalValidationError> {
+        // TODO
+        // ValSemXXX (1/2): Proposal list must not contain multiple PreSharedKey proposals that
+        //                  reference the same PreSharedKeyID.
+        let mut visited_psk_ids = Vec::new();
+
+        for proposal in proposal_queue.psk_proposals() {
+            let psk_id = proposal.psk_proposal().clone().into_psk_id();
+
+            match psk_id.psk() {
+                // TODO
+                // ValSemXXX: PSK in proposal must be of type Resumption/Application ...
+                Psk::Resumption(resumption_psk) => {
+                    if resumption_psk.usage != ResumptionPskUsage::Application {
+                        return Err(ProposalValidationError::Psk(PskError::UsageMismatch {
+                            allowed: vec![ResumptionPskUsage::Application],
+                            got: resumption_psk.usage,
+                        }));
+                    }
+                }
+                // ValSemXXX: ... or External.
+                Psk::External(_) => {}
+            };
+
+            // TODO
+            // ValSemXXX: The `psk_nonce` of a PreSharedKeyID MUST have length KDF.Nh.
+            {
+                let expected_nonce_length = self.ciphersuite().hash_length();
+                let got_nonce_length = psk_id.psk_nonce().len();
+
+                if expected_nonce_length != got_nonce_length {
+                    return Err(ProposalValidationError::Psk(
+                        PskError::NonceLengthMismatch {
+                            expected: expected_nonce_length,
+                            got: got_nonce_length,
+                        },
+                    ));
+                }
+            }
+
+            // TODO
+            // ValSemXXX (2/2): Proposal list must not contain multiple PreSharedKey proposals that
+            //                  reference the same PreSharedKeyID.
+            if !visited_psk_ids.contains(&psk_id) {
+                visited_psk_ids.push(psk_id);
```

Comment:
Ah so you're storing the entire PskId. Wouldn't a hash or something be enough?
Anyway, this is fine for now.